slides/module3.md

```diff
@@ -28,3 +34,573 @@ add the skills the trainee will obtain
 specify the agenda
 
 ------
+
+# Objects
+
+Objects in Scala are used for singletons and utility methods.
+
+```scala
+object MySingleton {
+  var n: Int = 3
+  def utilityMethod1(): String = ???
+}
+```
+
+**Other** languages have *static* methods or fields instead.
+
+------
+
+# Classes
+
+Scala supports OOP concepts, like classes:
+
+```scala
+// Definition
+class Person(val firstName: String,
+             val lastName: String,
+             var age: Int)
+
+// Usage
+val p = new Person("Clark", "Kent", 30)
+```
+
+. . .
+
+- Class names should be capitalized
+- Primary constructor is in the class signature
+
+Check [Classes tour].
+
+[Classes tour]: https://docs.scala-lang.org/tour/classes.html
+
+------
+
+# Companion Object
+
+An object of the same name than a class in the **same** source file
+
+```scala
+class Account
+
+// Companion object
+object Account {
+}
+```
+
+------
+
+# Case classes
+
+Case classes example
+
+```scala
+case class Person(firstName: String,
+                  lastName: String,
+                  age: Int)
+```
+
+------
+
+## Advantages over simple class
+
+- Optimized for pattern matching
+- Every parameter constructor becomes a `val`.
+- `apply` method automatically provided (no need to use `new`)
+- `unapply` method automatically provided (pattern matching)
+- Automatic `toString`, `equals`, `hashCode`, and `copy` unless explicitly provided.
+
+------
+
+### Simple pattern matching
+
+Programmers and Editors
+
+```scala
+sealed trait ProgrammingLanguage
+case object CSharp      extends ProgrammingLanguage
```

Comment:
Let's be careful about using language constructs that were not explained/exposed earlier.